### ISSUE TITLE
Inconsistant data retrieval between Localstorage and WebSQL

### ISSUE THREAD

**Issue by adamcbrewer** [mozilla/localForage#357]
#### Background
I've been pulled onto a project that runs as a questionnaire for an in-store, standalone app. The whole thing uses offline storage and appcache (which is a nightmare to debug) and so LocalForage is used to handle keeping data stored in the event that a device running the app goes offline. The app itself uses the the default LocalForage config.

#### The Error
Simply put, various browsers and devices use different storage APIs. Retrieving stored data returns the stringified store item object, but in the case of an he iPad 2 retina using WebSQL the returned data is already parsed and returned as an object.

Simply put, there should be a consistent returned value from the store. IndexDB and LocalStorage return strings and objects respectively. This could case unforeseen errors where the config isn't set to something specific.

**Comment by tofumatt** [comment]
Sorry, but localForage *should* always return the same object back across drivers. Can you should me a test case/code example where you get different results in WebSQL versus IndexedDB? Especially with basic types like String and Number, the return values should always be the same across drivers/browsers.

**Comment by adamcbrewer** [comment]
Thank's for replying @tofumatt. I realise now the version this was build with was version 1 – this probably had something to do with it.

I spent most of yesterday pushing out a quick patch, so it's not so easy to  provide some screenshots at the moment, but the code blow is pretty much what was used (excuse the nesting):

```js

var testData = [
    {
        test: true,
        someValue: 'string'
    }
];

localforage.setItem('testKey', JSON.stringify(testData), function () {

    localforage.getItem('testKey', function (data) {
        
        // fails when using WebSQL as 'data' is an object
        // and jQuery.parseJSON throws and error if 'data'
        // is not a string
        var dataParsedError = $.parseJSON(data);

        // Works fine as this method returns the original object
        // if passed something other that a string.
        var dataParsedSuccess = JSON.parse(data);

    });

});
```

**Comment by tofumatt** [comment]
Strange, we use JSON.parse internally but I think there could have been some errors with it early on.

Did upgrading to a newer version fix your issue?

-tofumatt

On 19 March 2015 at 10:42:31, Adam Brewer (notifications@github.com) wrote:

Thank's for replying @tofumatt. I realise now the version this was build with was version 1 – this probably had something to do with it.

I spent most of yesterday pushing out a quick patch, so it's not so easy to provide some screenshots at the moment, but the code blow is pretty much what was used (excuse the nesting):

var testData = [
    {
        test: true,
        someValue: 'string'
    }
];

localforage.setItem('testKey', JSON.stringify(testData), function () {

    localforage.getItem('testKey', function (data) {

        // fails when using WebSQL as 'data' is an object
        // and jQuery.parseJSON throws and error if 'data'
        // is not a string
        var dataParsedError = $.parseJSON(data);

        // Works fine as this method returns the original object
        // if passed something other that a string.
        var dataParsedSuccess = JSON.parse(data);

    });

});

**Comment by adamcbrewer** [comment]
A lot of the code that was written is using anonymous functions, not promises, so I'm waiting for a quiet period to try the upgrade. Without modifying the code an upgrade currently breaks the app.

I'm thinking this would be an obvious issue if the problem persisted to the current version. Given that the error happened with an older version do you think this isn't an issue anymore? Happy to close it if you agree.

**Comment by tofumatt** [comment]
I don’t think it’s an issue, and the new version actually still supports anonymous functions, so you should be able to upgrade without issues.

Sure, let’s close this unless you can reproduce it. Feel free to open it if you’re having the same issue with a newer version!

-tofumatt

On 20 March 2015 at 16:07:08, Adam Brewer (notifications@github.com) wrote:

A lot of the code that was written is using anonymous functions, not promises, so I'm waiting for a quiet period to try the upgrade. Without modifying the code an upgrade currently breaks the app.

I'm thinking this would be an obvious issue if the problem persisted to the current version. Given that the error happened with an older version do you think this isn't an issue anymore? Happy to close it if you agree.

**Comment by adamcbrewer** [comment]
Will do – thanks Matt!
